### ISSUE TITLE
Improve adjudicator allocation state signals/messaging

### ISSUE THREAD

**Issue by philipbelesky** [TabbycatDebate/tabbycat#357]
Superseding some issues in #258. Couple of things to think about:
- How to indicate auto saving is taking place? Errors in auto saving will throw a modal; but it may be unclear what is happening — or a connection may cut off
  - [X] Perhaps a small "Last saved at 3:42" text field somewhere that updates after successful POSTS
- How to warn users that changing allocations may lead to a loss of data when results/feedback exists?
  - [x] Perhaps a lock icon around those particular debates? For results at least the data is being pulled anyway
  - The server-side view that handles the POST could similarly do a check and bounce back a warning message in a modal; however what happens then? The state resets? Maybe part of the the POST could be a `confirmed` variable; where the modal message offers a `confirm change` button (> submit with `confirmed`) or a `cancel change` button (revert state)
- Provide better warnings on why an autoallocation may fail
  - [x] Test/enhance current messages to ensure they surface the reason why an auto allocation may fail

**Comment by philipbelesky** [comment]
Something worth considering: should we ever allow adjudicators to be removed from debates when a ballot exists? Just noticed a recent error email: `DoesNotExist at /t/NJDO2016/admin/results/round/1/ DebateAdjudicator matching query does not exist.` which seems to suggest that can happen in the wild. 

Presuming the change was made through the allocation interface that seems like a good place to explain that the ballot must be removed in order for the change to be made. Is there any common circumstance where you would or could not delete/re-renter the affected ballot in order to then make the allocation change?

**Comment by czlee** [comment]
Hmm. I guess in some circumstances adj-cores make on-the-fly adjudicator changes, because of unseen conflicts or the like. But one would like to think that it would be corrected before a ballot is entered, or at the very least, data entry folk would catch it and refuse to enter the ballot until the adjudicator had been corrected. I think post-draw-release adjudicator changes are common, but in order to get to the point where you need to change it _after_ the ballot's been entered, it has to slip through a lot of cracks.

If they've already discarded the paper ballot, users can (having been prevented from going ahead with the adj change) go back and copy down the numbers by hand, then re-enter them. If they're using e-ballots, with randomised URLs they can't enter it until adj allocations are fixed anyway; with public non-randomised URLs, well, maybe user error there might be an issue, but hopefully that's anomalous enough and adj cores can ask adjudicators not to enter results if adj allocations are wrong.

Bottom line is I'm comfortable enforcing this restriction in the UI.

Would you enforce this on a per-debate or whole-round basis?

**Comment by czlee** [comment]
Hmm, actually, I like to think that result.py would be robust enough to handle this problem more gracefully than that. I might look into it later…

**Comment by philipbelesky** [comment]
Not sure what the per-round vs per-debate basis would be. What I was imagining was essentially that the allocation UI would check for this scenario before saving the allocation change, and if it finds this case it would display a message to the user explaining what has happened and how to fix it.

**Comment by czlee** [comment]
Oh, after the fact, not before?

By per-round I meant the user would be prevented/warned about editing _any_ adjudicators once _any_ result is in for the round. By per-debate I meant the warning would only appear if a debate with a result was about to be affected.

**Comment by czlee** [comment]
Actually, I'm at a loss for how that NJDO2016 error is possible. It's not failing on result lookup—it's failing as it tries to retrieve a `DebateAdjudicator` instance using `debate` and an item in `debate.adjudicators.list`, which is meant to be just a list of adjudicators relating to related `DebateAdjudicator` instances. So although the error is in result.py, it's not consistent with behaviour you would expect if it were caused by results sticking around, or not sticking around, after an adj allocation was changed.

The only thing I can think of is a race condition, if somehow in the time between when it loaded the `BallotSet` and constituent `Scoresheet` objects, someone changed an adjudicator…

**Comment by philipbelesky** [comment]
Merging in #398 to this:

- [ ] The errors modified in the Hungarian algorithm code in d24330d should be surfaced to the user interface, so that after the fact users are alerted to the fact that the allocation did not work properly. This probably requires AJAX.
